src/commonMain/kotlin/io/konform/validation/ValidationBuilder.kt

```diff
@@ -1,70 +1,111 @@
     @JvmName("onEachArray")
-    infix fun <R> KProperty1<T, Array<R>>.onEach(init: ValidationBuilder<R>.() -> Unit) = onEachArray(this, init)
-    internal abstract fun <K, V> onEachMap(prop: KProperty1<T, Map<K, V>>, init: ValidationBuilder<Map.Entry<K, V>>.() -> Unit)
+    infix fun <R> KFunction1<T, Array<R>>.onEach(init: ValidationBuilder<C, R, E>.() -> Unit) = onEachArray(this.name, this, init)
+
+    internal abstract fun <K, V> onEachMap(name: String, mapFn: (T) -> Map<K, V>, init: ValidationBuilder<C, Map.Entry<K, V>, E>.() -> Unit)
+    @JvmName("onEachMap")
+    infix fun <K, V> KProperty1<T, Map<K, V>>.onEach(init: ValidationBuilder<C, Map.Entry<K, V>, E>.() -> Unit) = onEachMap(this.name, this, init)
     @JvmName("onEachMap")
-    infix fun <K, V> KProperty1<T, Map<K, V>>.onEach(init: ValidationBuilder<Map.Entry<K, V>>.() -> Unit) = onEachMap(this, init)
-    abstract infix fun <R> KProperty1<T, R?>.ifPresent(init: ValidationBuilder<R>.() -> Unit)
-    abstract infix fun <R> KProperty1<T, R?>.required(init: ValidationBuilder<R>.() -> Unit)
-    abstract fun run(validation: Validation<T>)
-    abstract val <R> KProperty1<T, R>.has: ValidationBuilder<R>
+    infix fun <K, V> KFunction1<T, Map<K, V>>.onEach(init: ValidationBuilder<C, Map.Entry<K, V>, E>.() -> Unit) = onEachMap(this.name, this, init)
+
+    internal abstract fun <R : Any> ifPresent(name: String, mapFn: (T) -> R?, init: ValidationBuilder<C, R, E>.() -> Unit)
+    infix fun <R : Any> KProperty1<T, R?>.ifPresent(init: ValidationBuilder<C, R, E>.() -> Unit) = ifPresent(this.name, this, init)
+    infix fun <R : Any> KFunction1<T, R?>.ifPresent(init: ValidationBuilder<C, R, E>.() -> Unit) = ifPresent(this.name, this, init)
+
+    internal abstract fun <R : Any> required(name: String, hint: HintBuilder<C, R?, E>, mapFn: (T) -> R?, init: ValidationBuilder<C, R, E>.() -> Unit): ConstraintBuilder<C, R?, E>
+    infix fun <R : Any> KProperty1<T, R?>.required(hintedBuilder: HintedRequiredBuilder<C, R, E>): ConstraintBuilder<C, R?, E> =
+        required(this.name, hintedBuilder.hint, this, hintedBuilder.init)
+    infix fun <R : Any> KFunction1<T, R?>.required(hintedBuilder: HintedRequiredBuilder<C, R, E>): ConstraintBuilder<C, R?, E> =
+        required(this.name, hintedBuilder.hint, this, hintedBuilder.init)
+
+    abstract fun <C, R, E> with(hint: HintBuilder<C, R?, E>, init: ValidationBuilder<C, R, E>.() -> Unit): HintedRequiredBuilder<C, R, E>
+    abstract fun <C, R> with(init: ValidationBuilder<C, R, String>.() -> Unit): HintedRequiredBuilder<C, R, String>
+
+    abstract fun <S> run(validation: Validation<S, T, E>, map: (C) -> S)
+    fun run(validation: Validation<C, T, E>) = run(validation, ::identity)
+
+    internal abstract fun add(builder: ComposableBuilder<C, T, E>)
+
+    abstract val <R> KProperty1<T, R>.has: ValidationBuilder<C, R, E>
+}
+
+interface ConstraintBuilder<C, T, E> {
+    infix fun hint(hint: HintBuilder<C, T, E>) : ConstraintBuilder<C, T, E>
```

Comment:
It should be possible to define an extension function to allow the old `hint "error message"` in addition to `hint stringHint("error message"`. Maybe `stringHint` could be `internal`.

```
infix fun <C,T> ConstraintBuilder<C,T,String>.hint(hint: String) = hint(stringHint(hint))
```

Reply:
Nice one!

I added this extension function and one on `ValidationBuilder`:
`fun <C, T> ValidationBuilder<C, T, String>.addConstraint(hint: String, vararg values: Any, test: C.(T) -> Boolean): ConstraintBuilder<C, T, String>` 

I would keep `stringHint` public. It provides the "templated" approach that was previously in the `NodeValidation`. I can imagine that it is very useful when creating your own constraint extensions.